src/XMonad/Layout.hs

```diff
@@ -230,16 +230,13 @@ instance (LayoutClass l a, LayoutClass r a) => LayoutClass (Choose l r) a where
     handleMessage c@(Choose d l r) m | Just ReleaseResources <- fromMessage m =
         join $ liftM2 (choose c d) (handle l ReleaseResources) (handle r ReleaseResources)
 
-    handleMessage c@(Choose d l r) m | Just (JumpToLayout desc) <- fromMessage m =
-        if | desc == description l -> choose c CL Nothing Nothing
-           | desc == description r -> choose c CR Nothing Nothing
-           | otherwise             -> case d of
-               CL -> handleMessage l m >>= \case
-                   Just ml -> choose c CL (Just ml) Nothing
-                   Nothing -> choose c CR Nothing =<< handleMessage r m
-               CR -> handleMessage r m >>= \case
-                   Just mr -> choose c CR Nothing (Just mr)
-                   Nothing -> (\ml -> choose c CL ml Nothing) =<< handleMessage l m
+    handleMessage c@(Choose d l r) m | Just (JumpToLayout desc) <- fromMessage m = do
+        ml <- handleMessage l m
+        mr <- handleMessage r m
```

Comment:
I should probably not even mention it to you, as we all know what's gonna happen, but still:

This simple implementation always walks the whole tree of Chooses and then possibly passes the message all the way down layout modifiers as well. The previous one stopped descending when it found a node with the desired description, but it still passed the message to some subtrees (all the way to layout modifiers and base layout). Asymptotically it's the same, in practice it might be a couple nanosecs or microsecs difference, I wouldn't worry about it at all. But it can be dealt with by keeping the first two lines of the multi-way if in here.

Reply:
I will go against expectations and prefer clarity to this spurious efficiency gain in this situation then :P